File: src/scala3/KDTree.scala
//> using scala "3.3.3"

package algorithms.scala3

final case class KDTree[A] private (k: Int, root: Option[KDTree.Node[A]], size: Int):
  import KDTree.{Entry, Node}

  def insert(point: Vector[Double], value: A): KDTree[A] =
    KDTree.validatePoint(k, point)
    KDTree.fromEntries(k, entries.updatedWith(point, value))

  def contains(point: Vector[Double]): Boolean =
    KDTree.validatePoint(k, point)
    def loop(node: Option[Node[A]]): Boolean =
      node match
        case None => false
        case Some(n) =>
          if n.entry.point == point then true
          else
            val axis = n.axis
            if point(axis) < n.entry.point(axis) then loop(n.left)
            else if point(axis) > n.entry.point(axis) then loop(n.right)
            // Equal on the split axis does not imply equal points, so both subtrees can still contain the target.
            else loop(n.left) || loop(n.right)
    loop(root)

  def get(point: Vector[Double]): Option[A] =
    KDTree.validatePoint(k, point)
    def loop(node: Option[Node[A]]): Option[A] =
      node match
        case None => None
        case Some(n) =>
          if n.entry.point == point then Some(n.entry.value)
          else
            val axis = n.axis
            if point(axis) < n.entry.point(axis) then loop(n.left)
            else if point(axis) > n.entry.point(axis) then loop(n.right)
            // Equal on this axis may still differ on another coordinate.
            else loop(n.left).orElse(loop(n.right))
    loop(root)

  def nearest(target: Vector[Double]): Option[Entry[A]] =
    KDTree.validatePoint(k, target)
    root.map { r =>
      val (bestEntry, _) =
        nearestNode(
          r,
          target,
          bestEntry = r.entry,
          bestDist2 = KDTree.dist2(r.entry.point, target)
        )
      bestEntry
    }

  def range(min: Vector[Double], max: Vector[Double]): Vector[Entry[A]] =
    KDTree.validatePoint(k, min)
    KDTree.validatePoint(k, max)
    def loop(node: Option[Node[A]], acc: Vector[Entry[A]]): Vector[Entry[A]] =
      node match
        case None => acc
        case Some(n) =>
          val axis = n.axis
          val inRange = KDTree.inRange(n.entry.point, min, max)
          val acc1 = if inRange then acc :+ n.entry else acc
          val acc2 = if min(axis) <= n.entry.point(axis) then loop(n.left, acc1) else acc1
          if max(axis) >= n.entry.point(axis) then loop(n.right, acc2) else acc2
    loop(root, Vector.empty)

  private def entries: Vector[Entry[A]] =
    def loop(node: Option[Node[A]], acc: Vector[Entry[A]]): Vector[Entry[A]] =
      node match
        case None => acc
        case Some(n) => loop(n.right, loop(n.left, acc :+ n.entry))
    loop(root, Vector.empty)

  private def nearestNode(node: Node[A], target: Vector[Double], bestEntry: Entry[A], bestDist2: Double): (Entry[A], Double) =
    val currentDist2 = KDTree.dist2(node.entry.point, target)
    val (newBestEntry, newBestDist2) =
      if currentDist2 < bestDist2 then (node.entry, currentDist2) else (bestEntry, bestDist2)

    val axis = node.axis
    val (near, far) =
      if target(axis) < node.entry.point(axis) then (node.left, node.right) else (node.right, node.left)

    val (bestAfterNearEntry, bestAfterNearDist2) =
      near match
        case None => (newBestEntry, newBestDist2)
        case Some(n) => nearestNode(n, target, newBestEntry, newBestDist2)

    // If the hypersphere around the target crosses the split plane, the far side may contain a closer point.
    val axisDelta = target(axis) - node.entry.point(axis)
    if axisDelta * axisDelta < bestAfterNearDist2 then
      far match
        case None => (bestAfterNearEntry, bestAfterNearDist2)
        case Some(n) => nearestNode(n, target, bestAfterNearEntry, bestAfterNearDist2)
    else
      (bestAfterNearEntry, bestAfterNearDist2)

object KDTree:
  final case class Entry[A](point: Vector[Double], value: A)
  final case class Node[A](entry: Entry[A], axis: Int, left: Option[Node[A]], right: Option[Node[A]])

  def empty[A](k: Int): KDTree[A] =
    require(k > 0, "k must be positive")
    KDTree[A](k, None, 0)

  def from[A](k: Int, values: Iterable[(Vector[Double], A)]): KDTree[A] =
    require(k > 0, "k must be positive")
    val entries =
      values.foldLeft(Vector.empty[Entry[A]]) { case (acc, (point, value)) =>
        validatePoint(k, point)
        acc.updatedWith(point, value)
      }
    fromEntries(k, entries)

  private def fromEntries[A](k: Int, entries: Vector[Entry[A]]): KDTree[A] =
    KDTree[A](k, build(entries, k, depth = 0), entries.size)

  private def build[A](entries: Vector[Entry[A]], k: Int, depth: Int): Option[Node[A]] =
    if entries.isEmpty then None
    else
      val axis = depth % k
      // Classic kd-tree construction: choose the median on the current axis to keep subtrees balanced.
      val sorted = entries.sortBy(_.point(axis))
      val median = sorted.size / 2
      Some(
        Node(
          entry = sorted(median),
          axis = axis,
          left = build(sorted.take(median), k, depth + 1),
          right = build(sorted.drop(median + 1), k, depth + 1)
        )
      )

  private def validatePoint(k: Int, point: Vector[Double]): Unit =
    require(point.size == k, s"Point dimension ${point.size} does not match k=$k")

  private def inRange(point: Vector[Double], min: Vector[Double], max: Vector[Double]): Boolean =
    var i = 0
    var ok = true
    while i < point.size && ok do
      ok = point(i) >= min(i) && point(i) <= max(i)
      i += 1
    ok

  private def dist2(a: Vector[Double], b: Vector[Double]): Double =
    var i = 0
    var sum = 0.0
    while i < a.size do
      val d = a(i) - b(i)
      // Squared distance avoids sqrt during comparisons; ordering is the same as Euclidean distance.
      sum += d * d
      i += 1
    sum

  extension [A](entries: Vector[Entry[A]])
    private def updatedWith(point: Vector[Double], value: A): Vector[Entry[A]] =
      val index = entries.indexWhere(_.point == point)
      val entry = Entry(point, value)
      if index >= 0 then entries.updated(index, entry) else entries :+ entry

File: src/scala3/KDTree.js
class KDTree {
  constructor(k, root = null, size = 0) {
    if (!Number.isInteger(k) || k <= 0) {
      throw new Error("k must be a positive integer");
    }

    this.k = k;
    this.root = root;
    this.size = size;
  }

  static empty(k) {
    return new KDTree(k);
  }

  static from(k, values) {
    if (!Number.isInteger(k) || k <= 0) {
      throw new Error("k must be a positive integer");
    }

    const entries = [];
    for (const [point, value] of values) {
      KDTree.#validatePoint(k, point);
      KDTree.#upsert(entries, point, value);
    }

    return KDTree.#fromEntries(k, entries);
  }

  insert(point, value) {
    KDTree.#validatePoint(this.k, point);
    const entries = this.#entries();
    KDTree.#upsert(entries, point, value);
    return KDTree.#fromEntries(this.k, entries);
  }

  contains(point) {
    KDTree.#validatePoint(this.k, point);

    const loop = (node) => {
      if (node === null) return false;
      if (KDTree.#samePoint(node.entry.point, point)) return true;

      const axis = node.axis;
      if (point[axis] < node.entry.point[axis]) return loop(node.left);
      if (point[axis] > node.entry.point[axis]) return loop(node.right);
      return loop(node.left) || loop(node.right);
    };

    return loop(this.root);
  }

  get(point) {
    KDTree.#validatePoint(this.k, point);

    const loop = (node) => {
      if (node === null) return undefined;
      if (KDTree.#samePoint(node.entry.point, point)) return node.entry.value;

      const axis = node.axis;
      if (point[axis] < node.entry.point[axis]) return loop(node.left);
      if (point[axis] > node.entry.point[axis]) return loop(node.right);

      const leftValue = loop(node.left);
      return leftValue === undefined ? loop(node.right) : leftValue;
    };

    return loop(this.root);
  }

  nearest(target) {
    KDTree.#validatePoint(this.k, target);
    if (this.root === null) return undefined;

    const best = this.#nearestNode(
      this.root,
      target,
      this.root.entry,
      KDTree.#dist2(this.root.entry.point, target)
    );

    return best.entry;
  }

  range(min, max) {
    KDTree.#validatePoint(this.k, min);
    KDTree.#validatePoint(this.k, max);

    const found = [];
    const loop = (node) => {
      if (node === null) return;

      const axis = node.axis;
      if (KDTree.#inRange(node.entry.point, min, max)) {
        found.push(node.entry);
      }
      if (min[axis] <= node.entry.point[axis]) loop(node.left);
      if (max[axis] >= node.entry.point[axis]) loop(node.right);
    };

    loop(this.root);
    return found;
  }

  #entries() {
    const entries = [];

    const loop = (node) => {
      if (node === null) return;
      entries.push(node.entry);
      loop(node.left);
      loop(node.right);
    };

    loop(this.root);
    return entries;
  }

  #nearestNode(node, target, bestEntry, bestDist2) {
    const currentDist2 = KDTree.#dist2(node.entry.point, target);
    let currentBestEntry = bestEntry;
    let currentBestDist2 = bestDist2;

    if (currentDist2 < bestDist2) {
      currentBestEntry = node.entry;
      currentBestDist2 = currentDist2;
    }

    const axis = node.axis;
    const near = target[axis] < node.entry.point[axis] ? node.left : node.right;
    const far = target[axis] < node.entry.point[axis] ? node.right : node.left;

    if (near !== null) {
      const bestAfterNear = this.#nearestNode(near, target, currentBestEntry, currentBestDist2);
      currentBestEntry = bestAfterNear.entry;
      currentBestDist2 = bestAfterNear.dist2;
    }

    const axisDelta = target[axis] - node.entry.point[axis];
    if (axisDelta * axisDelta < currentBestDist2 && far !== null) {
      return this.#nearestNode(far, target, currentBestEntry, currentBestDist2);
    }

    return { entry: currentBestEntry, dist2: currentBestDist2 };
  }

  static #fromEntries(k, entries) {
    return new KDTree(k, KDTree.#build(entries, k, 0), entries.length);
  }

  static #build(entries, k, depth) {
    if (entries.length === 0) return null;

    const axis = depth % k;
    const sorted = [...entries].sort((a, b) => a.point[axis] - b.point[axis]);
    const median = Math.floor(sorted.length / 2);

    return {
      entry: sorted[median],
      axis,
      left: KDTree.#build(sorted.slice(0, median), k, depth + 1),
      right: KDTree.#build(sorted.slice(median + 1), k, depth + 1),
    };
  }

  static #upsert(entries, point, value) {
    const index = entries.findIndex((entry) => KDTree.#samePoint(entry.point, point));
    const entry = { point: [...point], value };

    if (index >= 0) entries[index] = entry;
    else entries.push(entry);
  }

  static #validatePoint(k, point) {
    if (!Array.isArray(point) || point.length !== k) {
      throw new Error(`Point dimension ${Array.isArray(point) ? point.length : "invalid"} does not match k=${k}`);
    }

    if (!point.every((coordinate) => typeof coordinate === "number" && Number.isFinite(coordinate))) {
      throw new Error("Point coordinates must be finite numbers");
    }
  }

  static #samePoint(a, b) {
    return a.length === b.length && a.every((value, index) => value === b[index]);
  }

  static #inRange(point, min, max) {
    return point.every((value, index) => value >= min[index] && value <= max[index]);
  }

  static #dist2(a, b) {
    return a.reduce((sum, value, index) => {
      const delta = value - b[index];
      return sum + delta * delta;
    }, 0);
  }
}

module.exports = KDTree;
